[CSS] Fix nested setSelectorText() to prepend implicit selector

mdubet opened the pull request:
#### 85f439ac76ff4256c6a9e662e309e82b11da10f4
<pre>
[CSS] Fix nested setSelectorText() to prepend implicit selector
<a href="https://bugs.webkit.org/show_bug.cgi?id=280446">https://bugs.webkit.org/show_bug.cgi?id=280446</a>
<a href="https://rdar.apple.com/136791222">rdar://136791222</a>

Reviewed by Antti Koivisto.

Previously, CSSParserImpl was doing the implicit selector prepend.
However when using the CSSOM setSelectorText(), we don&apos;t use CSSParserImpl at all.

This patch:
  * adds more information in the NestedContext structure (kind of the closest ancestor rule)
  * moves the implicit selector preprend from CSSParserImpl to CSSSelectorParser
  * passes the NestedContext to the selector parser on setSelectorText()
  * cleanups some unused parameter or redundant members

* LayoutTests/fast/css/cssom-insertrule-crash-expected.html:

Fix wrong expectations (the &amp; selector should always be serialized)

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/nested-declarations-matching-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/nested-declarations-matching.html:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::insertRule):
* Source/WebCore/css/CSSRule.cpp:
(WebCore::CSSRule::parserContext const):
(WebCore::CSSRule::nestedContext const):
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::insertRule):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSupportsCondition):
* Source/WebCore/css/parser/CSSParser.h:
(WebCore::CSSParser::parseRule):
(WebCore::CSSParser::parseSelectorList):
* Source/WebCore/css/parser/CSSParserEnum.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::CSSParserImpl):
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::CSSParserImpl::consumeImportRule):
(WebCore::CSSParserImpl::consumeSupportsRule):
(WebCore::CSSParserImpl::consumeScopeRule):
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeBlockContent):
(WebCore::appendImplicitSelectorPseudoClassScopeIfNeeded): Deleted.
(WebCore::appendImplicitSelectorNestingParentIfNeeded): Deleted.
(WebCore::CSSParserImpl::appendImplicitSelectorIfNeeded): Deleted.
* Source/WebCore/css/parser/CSSParserImpl.h:
(WebCore::CSSParserImpl::CSSParserImpl):
(WebCore::CSSParserImpl::parseRule):
(WebCore::CSSParserImpl::isStyleNestedContext):
(WebCore::CSSParserImpl::lastAncestorRuleType const):
(WebCore::CSSParserImpl::isNestedContext): Deleted.

Remove unused m_scopeRuleNestingLevel (we only need m_styleRuleNestingLevel)

* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::appendImplicitSelectorPseudoClassScopeIfNeeded):
(WebCore::appendImplicitSelectorNestingParentIfNeeded):
(WebCore::appendImplicitSelectorIfNeeded):
(WebCore::parseMutableCSSSelectorList):
* Source/WebCore/css/parser/CSSSelectorParser.h:
(WebCore::CSSSelectorParser::CSSSelectorParser):
(WebCore::parseCSSSelectorList):
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::supportsCondition):
* Source/WebCore/css/parser/CSSSupportsParser.h:
(WebCore::CSSSupportsParser::CSSSupportsParser):

Remove unused isNestedContext parameter

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::isValidRuleHeaderText):
(WebCore::InspectorStyleSheet::setRuleHeaderText):
(WebCore::isNestedContext): Deleted.

Canonical link: <a href="https://commits.webkit.org/284537@main">https://commits.webkit.org/284537@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/862557891136aee5a5d33bafd37fcc3bf8478c1c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69731 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49131 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22484 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73816 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20889 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71848 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56932 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20740 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55391 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13853 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72797 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44784 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60137 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35871 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41449 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17577 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19266 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63383 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17922 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75529 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/13867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17177 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/13903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60220 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62997 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15487 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4608 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44935 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46009 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47280 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45750 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSRule.cpp:
```diff
@@ -70,6 +72,18 @@ bool CSSRule::hasStyleRuleAncestor() const
+        if (is<CSSStyleRule>(parentRule))
+            return CSSParserEnum::AncestorRuleType::Style;
+        if (is<CSSScopeRule>(parentRule))
+            return CSSParserEnum::AncestorRuleType::Scope;
```
```suggestion
        if (is<CSSStyleRule>(*parentRule))
            return CSSParserEnum::AncestorRuleType::Style;
        if (is<CSSScopeRule>(*parentRule))
            return CSSParserEnum::AncestorRuleType::Scope;
```

This reduces a null check I believe.

anttijk commented on Source/WebCore/css/parser/CSSParserEnum.h:
```diff
@@ -32,7 +32,12 @@ namespace WebCore {
+    Scope,
+};
+
+using IsNestedContext = std::optional<AncestorRuleType>;
```
Having typealias called `IsNestedContext` for `std::optional<AncestorRuleType>` is confusing to the point I don't understand what it is trying to say. Can you come up with a better factoring/naming here?

anttijk commented on Source/WebCore/css/CSSRule.cpp:
```diff
@@ -70,6 +72,18 @@ bool CSSRule::hasStyleRuleAncestor() const
+    }
+
+    return { };
+}
```
Traversing ancestors is not great but I suppose it is not a practical problem.

mdubet replied:
It's only happening on CSSOM indeed

anttijk reviewed (approved):
Ok, but naming requires some work.

mdubet committed:
[CSS] Fix nested setSelectorText() to prepend implicit selector
https://bugs.webkit.org/show_bug.cgi?id=280446
rdar://136791222

Reviewed by Antti Koivisto.

Previously, CSSParserImpl was doing the implicit selector prepend.
However when using the CSSOM setSelectorText(), we don't use CSSParserImpl at all.

This patch:
  * adds more information in the NestedContext structure (kind of the closest ancestor rule)
  * moves the implicit selector preprend from CSSParserImpl to CSSSelectorParser
  * passes the NestedContext to the selector parser on setSelectorText()
  * cleanups some unused parameter or redundant members

* LayoutTests/fast/css/cssom-insertrule-crash-expected.html:

Fix wrong expectations (the & selector should always be serialized)

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/nested-declarations-matching-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/nested-declarations-matching.html:
* Source/WebCore/css/CSSGroupingRule.cpp:
(WebCore::CSSGroupingRule::insertRule):
* Source/WebCore/css/CSSRule.cpp:
(WebCore::CSSRule::parserContext const):
(WebCore::CSSRule::nestedContext const):
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
(WebCore::CSSStyleRule::insertRule):
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSupportsCondition):
* Source/WebCore/css/parser/CSSParser.h:
(WebCore::CSSParser::parseRule):
(WebCore::CSSParser::parseSelectorList):
* Source/WebCore/css/parser/CSSParserEnum.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::CSSParserImpl):
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::CSSParserImpl::consumeImportRule):
(WebCore::CSSParserImpl::consumeSupportsRule):
(WebCore::CSSParserImpl::consumeScopeRule):
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeBlockContent):
(WebCore::appendImplicitSelectorPseudoClassScopeIfNeeded): Deleted.
(WebCore::appendImplicitSelectorNestingParentIfNeeded): Deleted.
(WebCore::CSSParserImpl::appendImplicitSelectorIfNeeded): Deleted.
* Source/WebCore/css/parser/CSSParserImpl.h:
(WebCore::CSSParserImpl::CSSParserImpl):
(WebCore::CSSParserImpl::parseRule):
(WebCore::CSSParserImpl::isStyleNestedContext):
(WebCore::CSSParserImpl::lastAncestorRuleType const):
(WebCore::CSSParserImpl::isNestedContext): Deleted.

Remove unused m_scopeRuleNestingLevel (we only need m_styleRuleNestingLevel)

* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::appendImplicitSelectorPseudoClassScopeIfNeeded):
(WebCore::appendImplicitSelectorNestingParentIfNeeded):
(WebCore::appendImplicitSelectorIfNeeded):
(WebCore::parseMutableCSSSelectorList):
* Source/WebCore/css/parser/CSSSelectorParser.h:
(WebCore::CSSSelectorParser::CSSSelectorParser):
(WebCore::parseCSSSelectorList):
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::supportsCondition):
* Source/WebCore/css/parser/CSSSupportsParser.h:
(WebCore::CSSSupportsParser::CSSSupportsParser):

Remove unused isNestedContext parameter

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::isValidRuleHeaderText):
(WebCore::InspectorStyleSheet::setRuleHeaderText):
(WebCore::isNestedContext): Deleted.

Canonical link: https://commits.webkit.org/284537@main
